src/graph.mojo

```diff
@@ -0,0 +1,163 @@
+    alias null_index = Node[DataType].null_index
+
+    # The list of nodes in the graph.
+    var _nodes: List[Node[DataType]]
+
+    # A mapping for random lookup of nodes by their mask.
+    # Used for slow lookup of nodes.
+    var _map: Dict[BitMask, Int]
+
+    fn __init__(inout self, owned first_value: DataType = Self.null_value):
+        """Initializes the graph.
+
+        Args:
+            first_value: The value stored in the first node,
+                         corresponding to an empty bitmask.
+        """
+        self._nodes = List[Node[DataType]]()
+        self._map = Dict[BitMask, Int]()
+        _ = self.add_node(BitMask(), first_value)
+
+    @always_inline
+    fn add_node(
+        inout self,
+        owned node_mask: BitMask,
+        owned value: DataType = Self.null_value,
+    ) -> Int:
+        """Adds a node to the graph.
+
+        Args:
+            node_mask: The mask of the node.
+            value:     The value stored in the node.
+
+        Returns:
+            The index of the added node.
+        """
+        self._map[node_mask] = len(self._nodes)
+        self._nodes.append(Node(node_mask, value))
+        return len(self._nodes) - 1
+
+    @always_inline
+    fn create_link(
+        inout self, from_node_index: Int, changed_bit: BitMask.IndexType
+    ) -> Int:
+        """Creates a link between two nodes.
+
+        Args:
+            from_node_index: The index of the node from which the link is created.
+            changed_bit:     The index of the bit that differs between the nodes.
+
+        Returns:
+            The index of the node to which the link is created.
+        """
+        new_mask = self._nodes[from_node_index].bit_mask
+        new_mask.flip(changed_bit)
+        optional_to_index = self._map.get(new_mask)
+        if optional_to_index:
+            to_node_index = optional_to_index.value()
+        else:
+            to_node_index = self.add_node(new_mask)
+
+        self._nodes[from_node_index].neighbours[
+            int(changed_bit)
+        ] = to_node_index
+        self._nodes[to_node_index].neighbours[
+            int(changed_bit)
+        ] = from_node_index
+
+        return to_node_index
+
+    fn get_node_index(
+        inout self,
+        start_node_index: Int = 0,
+        *different_bits: BitMask.IndexType,
+    ) -> Int:
+        """Returns the index of the node differing from the start node
+        by the given indices.
+
+        Args:
+            start_node_index: The index of the start node.
+            different_bits:   The indices of the bits that differ between the nodes.
+
+        Returns:
+            The index of the node differing from the start node by the given indices.
+        """
+        var current_node = start_node_index
+        for mask_index in different_bits:
+            var next_node = self._nodes[current_node].neighbours[
+                int(mask_index)
+            ]
+            if next_node == Self.null_index:
```

Comment:
Ah, this resolves my previous comment.

It would probably be good to document this, here and in `create_link`.

Reply:
Is it *necessary* to do this check? I think in the worst case, `create_link` would be less efficient than possible, but it will not be wrong even if it is called on an existing link.

Reply:
You are right. When commenting above, I had not yet seen the check here. So all fine, I would just document here "creates a new node and links it if not already present", and "does not check whether the link is already established" in `create_link`.

Reply:
Well, but even if I did *not* check it, would it be wrong? 
I am happy to include the comments, though!